From 75022bd1c35f7f23d1054c23094ef96df9952b28 Mon Sep 17 00:00:00 2001
From: ruslanbay <67730802+ruslanbay@users.noreply.github.com>
Date: Sat, 27 Dec 2025 12:48:58 +0100
Subject: [PATCH 23/54] media: intel/ipu4: use v4l2_get_link_freq() helper

Simplify the retrieval of the CSI-2 link frequency by using the
standard v4l2_get_link_freq() helper instead of manually
querying V4L2_CID_LINK_FREQ and its menu items.

Additionally, update the link_freq parameter type to s64 to
match the helper's return type and improve error reporting
in the debug logs.

Link: https://github.com/intel/ipu6-drivers/blob/master/drivers/media/pci/intel/ipu-isys-csi2.c#L84
---
 drivers/media/pci/intel/ipu-isys-csi2.c | 39 ++++++++++---------------
 1 file changed, 15 insertions(+), 24 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-isys-csi2.c b/drivers/media/pci/intel/ipu-isys-csi2.c
index 4896984c7ed2..dcc0e39e65dd 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2.c
@@ -90,44 +90,35 @@ static struct v4l2_subdev_internal_ops csi2_sd_internal_ops = {
 	.close = ipu_isys_subdev_close,
 };
 
-int ipu_isys_csi2_get_link_freq(struct ipu_isys_csi2 *csi2, __s64 *link_freq)
+int ipu_isys_csi2_get_link_freq(struct ipu_isys_csi2 *csi2, s64 *link_freq)
 {
-	struct ipu_isys_pipeline *pipe = container_of(media_entity_pipeline(&csi2->asd.sd.entity),
+	struct media_pipeline *mp = media_entity_pipeline(&csi2->asd.sd.entity);
+	struct ipu_isys_pipeline *pipe = container_of(mp,
 						      struct ipu_isys_pipeline,
 						      pipe);
 	struct v4l2_subdev *ext_sd =
-	    media_entity_to_v4l2_subdev(pipe->external->entity);
-	struct v4l2_ext_control c = {.id = V4L2_CID_LINK_FREQ, };
-	struct v4l2_ext_controls cs = {.count = 1,
-		.controls = &c,
-	};
-	struct v4l2_querymenu qm = {.id = c.id, };
-	int rval;
+		media_entity_to_v4l2_subdev(pipe->external->entity);
+	struct device *dev = &csi2->isys->adev->dev;
+	unsigned int bpp, lanes;
+	s64 ret;
 
 	if (!ext_sd) {
 		WARN_ON(1);
 		return -ENODEV;
 	}
-	rval = v4l2_g_ext_ctrls(ext_sd->ctrl_handler, &csi2->av_meta.vdev, &csi2->isys->media_dev, &cs);
-	if (rval) {
-		dev_info(&csi2->isys->adev->dev, "can't get link frequency\n");
-		return rval;
-	}
 
-	qm.index = c.value;
+	bpp = ipu_isys_mbus_code_to_bpp((*csi2->asd.ffmt)->code);
+	lanes = csi2->nlanes;
 
-	rval = v4l2_querymenu(ext_sd->ctrl_handler, &qm);
-	if (rval) {
-		dev_info(&csi2->isys->adev->dev, "can't get menu item\n");
-		return rval;
+	ret = v4l2_get_link_freq(ext_sd->ctrl_handler, bpp, lanes * 2);
+	if (ret < 0) {
+		dev_err(dev, "can't get link frequency (%lld)\n", ret);
+		return ret;
 	}
 
-	dev_dbg(&csi2->isys->adev->dev, "%s: link frequency %lld\n", __func__,
-		qm.value);
+	dev_dbg(dev, "link freq of %s is %lld\n", ext_sd->name, ret);
+	*link_freq = ret;
 
-	if (!qm.value)
-		return -EINVAL;
-	*link_freq = qm.value;
 	return 0;
 }
 
-- 
2.51.0

